File: top_interview_questions/python/289_game_of_life.py
#!/usr/bin/env python
# encoding: utf-8
'''
@author: Fizz Cao 
@file: 289_game_of_life.py
@time: 5/10/2021 12:11 am
'''
from typing import List


class Solution:
    def gameOfLife(self, board: List[List[int]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        rows = len(board)
        columns = len(board[0])

        neighbors = [(1, 0), (0, 1), (-1, 0), (0, -1), (1, 1), (-1, -1), (-1, 1), (1, -1)]

        for row in range(rows):
            for column in range(columns):
                live_neighbors = 0
                for neighbor in neighbors:
                    r = row + neighbor[0]
                    c = column + neighbor[1]
                    if r >= 0 and r < rows and c >= 0 and c < columns and abs(board[r][c]) == 1:
                        live_neighbors += 1
                if board[row][column] == 1 and (live_neighbors < 2 or live_neighbors > 3):
                    board[row][column] = -1
                if board[row][column] == 0 and (live_neighbors == 3):
                    board[row][column] = 2

        for row in range(rows):
            for column in range(columns):
                if (board[row][column] == -1):
                    board[row][column] = 0
                if (board[row][column] == 2):
                    board[row][column] = 1

File: top_interview_questions/java/289_game_of_life.java
class Solution {
    public void gameOfLife(int[][] board) {
        int rows =  board.length;
        int columns = board[0].length;

        int[][] neighbors = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}, {1, 1}, {-1, -1}, {-1, 1}, {1, -1}};

        for (int row = 0; row < rows; row++) {
            for (int column = 0; column < columns; column++) {
                int live_neighbors = 0;
                for (int[] neighbor: neighbors) {
                    int r = row + neighbor[0];
                    int c = column + neighbor[1];
                    if (r >= 0 && r < rows && c >= 0 && c < columns && Math.abs(board[r][c]) == 1)
                        live_neighbors += 1;
                }
                if (board[row][column] == 1 && (live_neighbors < 2 || live_neighbors > 3))
                    board[row][column] = -1;
                if (board[row][column] == 0 && live_neighbors == 3)
                    board[row][column] = 2;
            }
        }

        for (int row = 0; row < rows; row++)
            for (int column = 0; column < columns; column++) {
                if (board[row][column] == -1)
                    board[row][column] = 0;
                if (board[row][column] == 2)
                    board[row][column] = 1;
            }
    }
}